implementations/rust/ockam/ockam_transport_udp/src/router/udp_router.rs

```diff
@@ -0,0 +1,184 @@
+}
+
+impl UdpRouter {
+    /// Create and register a new UDP router with the node context
+    pub(crate) async fn register(ctx: &Context) -> Result<UdpRouterHandle> {
+        let main_addr = Address::random_local();
+        let api_addr = Address::random_local();
+
+        let child_ctx = ctx.new_detached(Address::random_local()).await?;
+
+        let router = Self {
+            ctx: child_ctx,
+            main_addr: main_addr.clone(),
+            api_addr: api_addr.clone(),
+            map: BTreeMap::new(),
+            allow_auto_connection: true,
+        };
+
+        let handle = router.create_self_handle(ctx).await?;
+
+        ctx.start_worker(vec![main_addr.clone(), api_addr], router)
+            .await?;
+        trace!("Registering UDP router for type = {}", crate::UDP);
+        ctx.register(crate::UDP, main_addr).await?;
+
+        Ok(handle)
+    }
+
+    /// Create a new `UdpRouterHandle` representing this router
+    async fn create_self_handle(&self, ctx: &Context) -> Result<UdpRouterHandle> {
+        let handle_ctx = ctx.new_detached(Address::random_local()).await?;
+        let handle = UdpRouterHandle::new(handle_ctx, self.api_addr.clone());
+        Ok(handle)
+    }
+
+    async fn handle_route(&mut self, ctx: &Context, mut msg: LocalMessage) -> Result<()> {
+        trace!(
+            "UDP route request: {:?}",
+            msg.transport().onward_route.next()
+        );
+
+        let onward = msg.transport().onward_route.next()?.clone();
+
+        let next = if let Some(n) = self.map.get(&onward) {
+            n.clone()
+        } else {
+            let peer_str = match String::from_utf8(onward.deref().clone()) {
+                Ok(s) => s,
+                Err(_e) => return Err(TransportError::UnknownRoute.into()),
+            };
+
+            if self.allow_auto_connection {
+                self.connect(peer_str).await?
+            } else {
+                return Err(TransportError::UnknownRoute.into());
+            }
+        };
+
+        let transport_msg = msg.transport_mut();
+        transport_msg.onward_route.step()?;
+        // Prepend peer socket addr so that sender can use it
+        transport_msg.onward_route.modify().prepend(onward);
+        transport_msg.onward_route.modify().prepend(next.clone());
+
+        ctx.send(next.clone(), msg).await?;
+
+        Ok(())
+    }
+
+    async fn handle_register(&mut self, accepts: Vec<Address>, self_addr: Address) -> Result<()> {
+        if let Some(f) = accepts.first().cloned() {
+            trace!("UDP registration request: {} => {}", f, self_addr);
+        } else {
+            error!("Tried to register a new client without passing any `Address`");
+            return Err(TransportError::InvalidAddress.into());
+        }
+
+        for accept in &accepts {
+            if self.map.contains_key(accept) {
+                // TODO: is returning OK right if addr(s) are already registered
+                return Ok(());
+            }
```

Comment:
In `ockam_transport_tcp` we do this:

```rust
for accept in &accepts {
    if self.map.contains_key(accept) {
        error!("TCP registration request failed, this address is already connected: {}", accept);
        return Err(TransportError::AlreadyConnected.into());
    }
}
```

Reply:
Thanks for pointing this out. A few points I want to clarify regarding how we want the UDP transport to handle registering. In my current implementation, the `UdpListenProcessor` sends a `UdpRouterMessage::Register` to the router for every `TransportMessage` it receives. If I make the router throw an error here, when a peer sends multiple messages, the router will throw errors after the node receives the first message. I guess my questions are,
1. do we want to let the listener keeps track of registered addr(s) so that it won't send as many register message to the router?
1. when is proper for the router to un-register a peer's addr(s) from its map?